internal/dataplane/parser/translate_httproute.go

```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
+			hostnamedRoutes = append(hostnamedRoutes, *route)
+			continue
+		}
+		for _, hostname := range route.Spec.Hostnames {
+			hostnamedRoute := route.DeepCopy()
+			hostnamedRoute.Spec.Hostnames = []gatewayv1beta1.Hostname{hostname}
+			hostnamedRoutes = append(hostnamedRoutes, *hostnamedRoute)
+		}
+	}
+	for _, route := range hostnamedRoutes {
+		for _, rule := range route.Spec.Rules {
+			for _, match := range rule.Matches {
+				newRoute := route.DeepCopy()
+				newRoute.Spec.Rules = []gatewayv1beta1.HTTPRouteRule{
+					{
+						Matches: []gatewayv1beta1.HTTPRouteMatch{
+							match,
+						},
+						BackendRefs: rule.BackendRefs,
+						Filters:     rule.Filters,
+					},
+				}
+				newHTTPRoutes = append(newHTTPRoutes, *newRoute)
+			}
+		}
+	}
+
+	return newHTTPRoutes
+}
+
+func sortHTTPRoutes(routes []gatewayv1beta1.HTTPRoute) {
+	sort.SliceStable(routes, func(i, j int) bool {
+		hasHostnamei := hasHostnames(routes[i])
+		hasHostnamej := hasHostnames(routes[j])
+
+		matchi := routes[i].Spec.Rules[0].Matches[0]
+		matchj := routes[j].Spec.Rules[0].Matches[0]
+		pathMatchingi := matchi.Path
+		pathMatchingj := matchj.Path
+		isExactPathMatchingi := isPathMatchingExact(pathMatchingi)
+		isExactPathMatchingj := isPathMatchingExact(pathMatchingj)
+		isPrefixPathMatchingi := isPathMatchingPrefix(pathMatchingi)
+		isPrefixPathMatchingj := isPathMatchingPrefix(pathMatchingj)
+
+		namespacedNamei := fmt.Sprintf("%s/%s", routes[i].Namespace, routes[i].Name)
+		namespacedNamej := fmt.Sprintf("%s/%s", routes[j].Namespace, routes[j].Name)
+
+		switch {
+		// hostname matching
+		case hasHostnamei && !hasHostnamej:
+			return true
+		case !hasHostnamei && hasHostnamej:
+			return false
+		case hasHostnamei && hasHostnamej && len(routes[i].Spec.Hostnames[0]) != len(routes[j].Spec.Hostnames[0]):
+			return len(routes[i].Spec.Hostnames[0]) > len(routes[j].Spec.Hostnames[0])
+
+		// exact path matching
+		case isExactPathMatchingi && !isExactPathMatchingj:
+			return true
+		case !isExactPathMatchingi && isExactPathMatchingj:
+			return false
+		case isExactPathMatchingi && isExactPathMatchingj && len(*pathMatchingi.Value) != len(*pathMatchingj.Value):
+			return len(*pathMatchingi.Value) > len(*pathMatchingj.Value)
+
+		// prefix path matching
+		case isPrefixPathMatchingi && !isPrefixPathMatchingj:
+			return true
+		case !isPrefixPathMatchingi && isPrefixPathMatchingj:
+			return false
+		case isPrefixPathMatchingi && isPrefixPathMatchingj && len(*pathMatchingi.Value) != len(*pathMatchingj.Value):
+			return len(*pathMatchingi.Value) > len(*pathMatchingj.Value)
+
+		// header matching
+		case len(matchi.Headers) != len(matchj.Headers):
+			return len(matchi.Headers) > len(matchj.Headers)
+
+		// queryparam matching
+		case len(matchi.QueryParams) != len(matchj.QueryParams):
+			return len(matchi.QueryParams) > len(matchj.QueryParams)
+
+		// creation timestamp
+		case routes[i].CreationTimestamp.Compare(routes[j].CreationTimestamp.Time) != 0:
```

Comment:
**Oldest** route should take the highest precedence, so here we should use `time.Before`.

Reply:
good catch